### PR TITLE
Phase 6: Stochastic Depth (DropPath) — Implicit Ensemble Regularization

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Transolver's 3 blocks have no stochastic regularization. Stochastic Depth (DropPath) randomly skips entire residual branches during training, creating an implicit ensemble of subnetworks of varying depth. This forces the model to learn more **redundant, robust representations** that don't rely on any single block — a form of structural regularization proven to improve OOD generalization in ViT, DeiT, and Swin Transformer.

**Why this might help p_tan especially:** The tandem foil OOD is the hardest case (p_tan=30.29 vs p_in=13.03). The model may be overfitting to structural patterns in single-foil geometry that break under tandem interference. DropPath prevents over-reliance on specific block computation paths, encouraging more geometry-agnostic representations.

**Paper:** Huang et al., "Deep Networks with Stochastic Depth", ECCV 2016. Used in DeiT (Touvron 2021, best drop_path_rate=0.1 for small models) and virtually all modern ViT variants. Not in our dead-ends list.

## Instructions

Add a `DropPath` module and apply it in the Transolver blocks. The change is ~20 lines in `train.py`.

**Step 1: Add DropPath class** (add near other utility classes, before the `Physics_Attention_Structured_Mesh` definition):

```python
class DropPath(nn.Module):
    """Stochastic Depth (Huang et al., ECCV 2016). Drops entire residual paths."""
    def __init__(self, drop_prob: float = 0.0):
        super().__init__()
        self.drop_prob = drop_prob

    def forward(self, x):
        if not self.training or self.drop_prob == 0.0:
            return x
        keep_prob = 1.0 - self.drop_prob
        shape = (x.shape[0],) + (1,) * (x.ndim - 1)
        rand_t = torch.rand(shape, dtype=x.dtype, device=x.device)
        rand_t = torch.floor(rand_t + keep_prob)
        return x * rand_t / keep_prob
```

**Step 2: Add `--drop_path_rate` argument** to the argparse section:

```python
parser.add_argument('--drop_path_rate', type=float, default=0.0,
                    help='Stochastic depth drop path rate (0=disabled, try 0.1)')
```

**Step 3: Pass drop path rates to TransolverBlock** using a linear schedule (deeper layers drop more):

In the `Transolver` model `__init__`, where TransolverBlocks are created, compute per-layer rates and pass them:

```python
# Compute per-layer drop path rates (linear schedule, layer 0 gets lowest rate)
dpr = [cfg.drop_path_rate * i / max(cfg.n_layers - 1, 1) for i in range(cfg.n_layers)]
# For n_layers=3 and drop_path_rate=0.1: dpr = [0.0, 0.05, 0.1]
```

Pass `dpr[i]` to each `TransolverBlock` constructor. In `TransolverBlock.__init__`:

```python
self.drop_path = DropPath(drop_prob) if drop_prob > 0.0 else nn.Identity()
```

In `TransolverBlock.forward()`, apply to both attention and FFN residuals:

```python
# BEFORE: x = x + self.attn(self.norm1(x), ...)
# AFTER:
x = x + self.drop_path(self.attn(self.norm1(x), ...))
# Similarly for FFN:
x = x + self.drop_path(self.ffn(self.norm2(x)))
```

**Experiment runs:** Try two drop rates with two seeds each (run all 4 sequentially):

```bash
# Rate 0.1, seeds 42 and 73
python train.py --agent askeladd --wandb_name "askeladd/dp01-seed42" \
  --wandb_group phase6/stochastic-depth --drop_path_rate 0.1 --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3

python train.py --agent askeladd --wandb_name "askeladd/dp01-seed73" \
  --wandb_group phase6/stochastic-depth --drop_path_rate 0.1 --seed 73 \
  [same flags as above]

python train.py --agent askeladd --wandb_name "askeladd/dp02-seed42" \
  --wandb_group phase6/stochastic-depth --drop_path_rate 0.2 --seed 42 \
  [same flags]

python train.py --agent askeladd --wandb_name "askeladd/dp02-seed73" \
  --wandb_group phase6/stochastic-depth --drop_path_rate 0.2 --seed 73 \
  [same flags]
```

W&B group: `phase6/stochastic-depth`

## Baseline

Current best single-model metrics (8-seed mean, seeds 66-73):
| Metric | Baseline |
|--------|----------|
| p_in | 13.03 |
| p_oodc | 7.83 |
| p_tan | 30.29 |
| p_re | 6.45 |

Individual seed references:
- seed=42: p_in≈12.71, p_oodc≈8.13
- seed=73: p_in≈12.2, p_oodc≈7.59

Ensemble (8 seeds 66-73): p_in=12.2, p_oodc=6.7, p_tan=29.1, p_re=5.8

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent askeladd --wandb_name "askeladd/baseline-seed42" \
  --seed 42 --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT askeladd:

## Results

### Individual Run Metrics

| Config | Seed | val/loss | p_in | p_oodc | p_tan | p_re | W&B Run |
|--------|------|----------|------|--------|-------|------|---------|
| DropPath 0.1 | 42 | 0.3965 | 13.3 | 8.2 | 30.7 | 6.6 | s93o0li2 |
| DropPath 0.1 | 73 | 0.4037 | 13.7 | 8.9 | 31.3 | 6.7 | fovffeeg |
| DropPath 0.2 | 42 | 0.4110 | 13.6 | 8.4 | 31.5 | 6.9 | 6dt5ofgn |
| DropPath 0.2 | 73 | 0.4033 | 13.0 | 8.8 | 30.9 | 6.8 | lqq102pu |

### Mean by Config

| Config | val/loss | p_in | p_oodc | p_tan | p_re |
|--------|----------|------|--------|-------|------|
| **Baseline** (8-seed mean) | **0.3829** | **13.03** | **7.83** | **30.29** | **6.45** |
| DropPath 0.1 (2-seed) | 0.4001 | 13.50 | 8.55 | 31.00 | 6.65 |
| DropPath 0.2 (2-seed) | 0.4072 | 13.30 | 8.60 | 31.20 | 6.85 |

### Per-Seed Comparison vs Known Baselines

The PR provides individual seed references. Comparing same-seed:

| Metric | Baseline s42 | DP 0.1 s42 | DP 0.2 s42 |
|--------|-------------|------------|------------|
| p_in | ~12.7¹ | 13.3 | 13.6 |
| p_oodc | ~8.1¹ | 8.2 | 8.4 |

| Metric | Baseline s73 | DP 0.1 s73 | DP 0.2 s73 |
|--------|-------------|------------|------------|
| p_in | ~12.2¹ | 13.7 | 13.0 |
| p_oodc | ~7.6¹ | 8.9 | 8.8 |

¹ Reference values from PR body

### Implementation Details

Added `DropPath` class with linear schedule across 3 layers:
- `drop_path_rate=0.1`: layer rates = [0.0, 0.05, 0.1]
- `drop_path_rate=0.2`: layer rates = [0.0, 0.1, 0.2]

Applied to both attention and FFN residual paths in TransolverBlock. ~20 lines of code added.

### Commands Used

```bash
# DropPath 0.1
CUDA_VISIBLE_DEVICES={0,1} python train.py --agent askeladd \
  --wandb_group phase6/stochastic-depth --drop_path_rate 0.1 --seed {42,73} \
  --wandb_name "askeladd/dp01-seed{42,73}" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3

# DropPath 0.2
CUDA_VISIBLE_DEVICES={2,3} python train.py --agent askeladd \
  --wandb_group phase6/stochastic-depth --drop_path_rate 0.2 --seed {42,73} \
  --wandb_name "askeladd/dp02-seed{42,73}" \
  [same flags as above]
```

W&B group: `phase6/stochastic-depth`

### What Happened

**Stochastic Depth hurts across the board.** Both drop rates degrade all metrics compared to the baseline:

- **p_oodc is the worst-hit metric** (+9-14% vs baseline) — the OOD condition split suffers most from dropping attention/FFN paths during training
- **p_tan shows modest degradation** (+2-3%) — contrary to the hypothesis, DropPath does NOT improve tandem transfer generalization
- **Higher drop rate (0.2) is slightly worse** than 0.1 on most metrics, though not dramatically
- **val/loss is significantly worse** (0.400-0.411 vs 0.383 baseline)

**Why it fails for this architecture:** Transolver has only 3 blocks — this is much shallower than typical ViT models where stochastic depth shines (12-24 layers in DeiT/Swin). With only 3 layers and a linear schedule, only the last block gets meaningful drop probability. The model can't afford to skip any of its limited computation capacity. The "implicit ensemble of subnetworks" benefit requires enough depth for meaningful path diversity.

### Suggested Follow-ups

- **DropPath is not suitable for 3-layer Transolver.** The architecture is too shallow for stochastic depth to provide its regularization benefits.
- If stochastic depth is revisited, it should only be tested if the model grows to 6+ layers.
- For OOD regularization, other approaches may work better: increased weight decay, label smoothing, or mixup/cutmix augmentation — these don't rely on architectural depth.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2099 because Closing as a dead end. W&B-confirmed results: all four DropPath runs are significantly worse than baseline across every metric (best DropPath 0.1 seed 42 gives p_in=13.32, p_oodc=8.19, p_tan=30.74, p_re=6.63 vs baseline p_in=12.1, p_oodc=6.6, p_tan=29.1, p_re=5.8). The student's diagnosis is correct: Transolver's 3-block architecture is too shallow for stochastic depth regularization to provide meaningful path diversity. DropPath shines in 12-24 layer ViTs — here it just removes compute capacity the model needs. Well-executed experiment with clear conclusion: revisit only if model depth reaches 6+ blocks..
